### PR TITLE
Facenet et al

### PR DESCRIPTION
**Pull request by ashish1595**

### CODE REVIEW

File: face-recognition/src/recognize.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                # perform classification to recognize the face
+                preds = recognizer.predict_proba(vec)[0]
+                j = np.argmax(preds)
+                probability = preds[j]
+                name = le.classes_[j]
+
+        # show the output image
+        return name, probability * 100
+    except Exception as ex:
+        print(str(ex))
```
Comment (rajan123456) [inline face-recognition/src/recognize.py:83]: Log messages using logger instead of print to console.

File: face-recognition/src/recognize.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                # face
+                box = detections[0, 0, i, 3:7] * np.array([w, h, w, h])
+                (startX, startY, endX, endY) = box.astype("int")
+
+                # extract_embeddings the face ROI
+                face = image[startY:endY, startX:endX]
+                (fH, fW) = face.shape[:2]
+
+                # ensure the face width and height are sufficiently large
+                if fW < 20 or fH < 20:
```
Comment (rajan123456) [inline face-recognition/src/recognize.py:63]: Read constants from configuration.

File: face-recognition/src/recognize.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                    continue
+
+                # construct a blob for the face ROI, then pass the blob
+                # through our face embedding model to obtain the 128-d
+                # quantification of the face
+                face_blob = cv2.dnn.blobFromImage(face, 1.0 / 255, (96, 96),
```
Comment (rajan123456) [inline face-recognition/src/recognize.py:69]: Read constants from configuration.

File: scripts/dev/docker-compose.yml [hunk 4]
```diff
@@ -56,7 +56,21 @@ services:
     ports:
       - "3000:80"
     depends_on:
-      - nginx
+      - user
```
Comment (rajan123456) [inline scripts/dev/docker-compose.yml:59]: The admin container dependencies should not be changed and should depend on Nginx for its startup.

File: scripts/dev/docker-compose.yml [hunk 5]
```diff
@@ -56,7 +56,21 @@ services:
...
+  facenet:
+    image: rajagupt/facenet:latest
+    container_name: facenet
+    restart: always
+    ports:
+      - "7000:5000"
+    volumes:
+      - ./ml-data-vol:/app/resources:ro
+  training:
```
Comment (rajan123456) [inline scripts/dev/docker-compose.yml:68]: Rename the component to watchdog

File: scripts/dev/docker-compose.yml [hunk 6]
```diff
@@ -304,3 +318,4 @@ volumes:
   vault-data-vol:
   zookeeper-data-vol:
   zookeeper-logs-vol:
+  ml-data-vol:
```
Comment (rajan123456) [inline scripts/dev/docker-compose.yml:321]: Use **facenet-data-vol** instead for the name of Facenet's volume.

File: training/src/constants.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# ELASTIC_APM = {
+#     'SERVICE_NAME': 'attendance-service',
+#     'SERVER_URL': 'http://apm-server:8200'
+# }
```
Comment (rajan123456) [inline training/src/constants.py:4]: Remove commented config entries for ELASTIC_APM which are not needed.

File: training/src/embeddings.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+import constants
+import logging
+
+log = logging.getLogger('root')
+
+def extract_embeddings():
+    log.info("embeddings extraction started.")
+    embeddings = constants.PICKLE_FILES_DIR + '/embeddings.pickle'
+    embedding_model = constants.MODEL_FILES_DIR + '/openface_nn4.small2.v1.t7'
+    conf = 0.5
```
Comment (rajan123456) [inline training/src/embeddings.py:17]: Read constants from configuration.

File: training/src/embeddings.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                    # the face
+                    box = detections[0, 0, i, 3:7] * np.array([w, h, w, h])
+                    (startX, startY, endX, endY) = box.astype("int")
+
+                    # extract_embeddings the face ROI and grab the ROI dimensions
+                    face = image[startY:endY, startX:endX]
+                    (fH, fW) = face.shape[:2]
+
+                    # ensure the face width and height are sufficiently large
+                    if fW < 20 or fH < 20:
```
Comment (rajan123456) [inline training/src/embeddings.py:85]: Read constants from configuration.

File: training/src/training_dataset_watchdog.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+def job():
+    already_trained_students_count = -1
+    students_image_count_obj = {}
+    try:
+        for username in os.listdir(constants.DATASET_PATH):
+            concatenated_dir = constants.DATASET_PATH + '/' + username
+            if os.path.isdir(concatenated_dir):
+                students_image_count_obj[username] = len([file for file in os.listdir(concatenated_dir) if (
+                            os.path.isfile(concatenated_dir + "/" + file) and os.path.splitext(file)[
+                        1].lower() == ".png")])
```
Comment (rajan123456) [inline training/src/training_dataset_watchdog.py:27]: Does streamProcessor generate images of png type?
Can we move this to supportedFileTypes property in the config?

Comment (anchitseth) [reply training/src/training_dataset_watchdog.py:27]: Yes, Images generated by streamProcessor are of .png extension.
supportedFileTypes -> do you want it to be a list of images extension or just png?

Comment (rajan123456) [reply training/src/training_dataset_watchdog.py:27]: Cool. For now, we can add a config entry **supportedFileType** with a single value **png**

**Pull request merged**